src/marked.js

```diff
@@ -139,44 +139,84 @@ marked.defaults = defaults;
  */
 
 marked.use = function(extension) {
-  const opts = merge({}, extension);
-  if (extension.renderer) {
-    const renderer = marked.defaults.renderer || new Renderer();
-    for (const prop in extension.renderer) {
-      const prevRenderer = renderer[prop];
-      renderer[prop] = (...args) => {
-        let ret = extension.renderer[prop].apply(renderer, args);
-        if (ret === false) {
-          ret = prevRenderer.apply(renderer, args);
+  if (!Array.isArray(extension)) { // Wrap in array if not already to unify processing
+    extension = [extension];
+  }
+  const opts = merge({}, ...extension);
+  opts.tokenizer = null;
+  opts.renderer = null;
+  const extensions = {};
+
+  extension.forEach((ext) => {
+    //= =-- Parse "addon" extensions --==//
+    if (ext.renderer && ext.name) { // Renderers must have 'name' property
+      extensions[ext.name] = ext.renderer;
+    }
+    if (ext.tokenizer && ext.level) { // Tokenizers must have 'level' property
+      if (extensions[ext.level]) {
+        extensions[ext.level].push(ext.tokenizer);
+      } else {
+        extensions[ext.level] = [ext.tokenizer];
+      }
+      if (ext.start) { // Regex to check for start of token
+        if (ext.level === 'block') {
+          extensions.startBlock = extensions.startBlock
+            ? new RegExp(extensions.startBlock.source + '|' + ext.start.source)
+            : ext.start;
+        } else if (ext.level === 'inline') {
+          extensions.startInline = extensions.startInline
+            ? new RegExp(extensions.startInline.source + '|' + ext.start.source)
```

Comment:
We will have to make sure this isn't a source of vulnerabilities. I'm kind of thinking `extensions.startInline` should be an array of functions/regexps. It might be a little slower if there are many extensions but it seems like it would be more secure and we aren't really guaranteeing speed when using extensions.

Reply:
and it would allow using means other than regexp to find the potential start

Reply:
Any guidelines on implementing this so it's secure? I'm not sure how separately calling regexes from an array is more secure than `or`ing them together or if there's a certain approach that would be more secure.

How much do we need to worry about security on user-made extensions like these in general? Any time the dev is constructing their own regexes like this, there's risk of exposing their own app to redos or whatever, and at some point that's out of our control.

Reply:
Also how are you envisioning this working with non-regex? Can you help me with examples so I can implement this correctly?

Reply:
> I'm not sure how separately calling regexes from an array is more secure than `or`ing them together or if there's a certain approach that would be more secure.

There are regexps that are fine separately but redos when together

For Example: `/(a*)*/` and `/b/` do not backtrack when run separately but `/(a*)*b/` does backtrack.

I don't know if there are regexps that are insecure when appended together with `|` but I also don't know that there aren't any.

> Any time the dev is constructing their own regexes like this, there's risk of exposing their own app to redos or whatever

The devs aren't "constructing their own regexes like this", that is part of the problem. If a dev uses the following code:

```js
const marked = require('marked')
const markedExtension1 = require('marked-extension-1')
const markedExtension2 = require('marked-extension-2')
marked.use([markedExtension1, markedExtension2])
```

The user might not even know that a regexp is being constructed when using those extensions. Also, as shown above, the extension authors might not be creating a vulnerable regexp but the act of combining them creates one by accident.

Reply:
> Also how are you envisioning this working with non-regex?

Any function that takes the `src` and returns an index of the first possible start. Something like:

```
start: (src) => src.indexOf(":")
```

Reply:
Ok that makes sense. I'll see if I can get that working. Should we just require `start` to be a function then to simplify our implementation? Then users can choose to use regex or string methods internally and we don't have to be type-checking their input?

Reply:
Ya that sounds good

Reply:
Done.